### ISSUE TITLE
Release V2

### ISSUE THREAD

**Issue by SeanNaren** [SeanNaren/deepspeech.pytorch#446]
Few improvements incoming. Currently waiting for models to train, and let me know if anyone is willing to help train baseline models!

Improvements

- Remove TorchAudio and use Scipy when loading audio based on speed comparisons and ease of installation
- Improved implementation of Nvidia Apex to make mixed precision training easier to use
- New pre-trained models using mixed-precision
- Documentation and improvements on how to tune and use librispeech LMs, and results based with the 4gram model

The changes for this can be seen on the V2 branch for anyone curious.

**Comment by igormq** [comment]
I can help you. I have 4 1080 and 2 1080TI in my lab :)

**Comment by SeanNaren** [comment]
Woop, thanks @igormq! Everything to run things is currently on a branch [here](https://github.com/SeanNaren/deepspeech.pytorch/tree/feature/V2). Here are a list of pre-trained models, the commands I plan to use to run them. I've already done the librispeech model, so now need a TEDlium + AN4 model to with them:

LibriSpeech Command (Model Done):
```
python train.py --rnn-type lstm --hidden-size 1024 --hidden-layers 5  --train-manifest data/libri_train_manifest.csv --val-manifest data/libri_val_manifest.csv --epochs 60 --num-workers 16 --cuda  --learning-anneal 1.01 --batch-size 64 --no-sortaGrad --visdom  --opt-level O1 --loss-scale 1 --id mixed_precision_lstm_libri_60_1.01 --checkpoint --save-folder deepspeech.pytorch/librispeech/ --model-path deepspeech.pytorch/librispeech/deepspeech_final.pth
```

TEDLium Command (**Not Started**):
```
python train.py --rnn-type lstm --hidden-size 1024 --hidden-layers 5  --train-manifest data/libri_train_manifest.csv --val-manifest data/libri_val_manifest.csv --epochs 60 --num-workers 16 --cuda  --learning-anneal 1.01 --batch-size 64 --no-sortaGrad --visdom  --opt-level O1 --loss-scale 1 --id mixed_precision_lstm_libri_60_1.01 --checkpoint --save-folder deepspeech.pytorch/tedlium/ --model-path deepspeech.pytorch/tedlium/deepspeech_final.pth
```

AN4 Command (**Not Started**):
```
python train.py --rnn-type lstm --hidden-size 1024 --hidden-layers 5  --train-manifest data/an4_train_manifest.csv --val-manifest data/an4_val_manifest.csv --epochs 70 --num-workers 16 --cuda  --learning-anneal 1.01 --batch-size 32 --no-sortaGrad --visdom  --opt-level O1 --loss-scale 1 --id mixed_precision_lstm_libri_60_1.01 --checkpoint --save-folder deepspeech.pytorch/an4/ --model-path deepspeech.pytorch/an4/deepspeech_final.pth
```

A few things:

- I've increased the model size and swapped to using LSTMs for convergence purposes. I've also seen better results using LSTMs and a larger model. I kept this standard across the datasets to reduce complexity.

- Parameters are based either of pre-existing parameters or my local setup, so you may need to play with the number of epochs (`--epochs`) and batch size (`--batch-size`) to fit your machine and get a model trained to convergence

The LibriSpeech model I trained does not perform better than the previous release, but the difference is marginal and the improvements to the library (in addition once documentation on how to use the LibriSpeech LM/Tune the Librispeech LM) will outperform the difference in accuracy!

**Comment by igormq** [comment]
Ok, I will start training the DeepSpeech model in the TEDLium Dataset =)

**Comment by madhavajay** [comment]
@SeanNaren Awesome! Out of interest what was the training time for the LibriSpeech dataset?

Are you able to share the model now, I want to set this up and confirm the inference time is acceptable on my target hardware before going through the process of training / finetuning.

Really excited about this!! :)

**Comment by SeanNaren** [comment]
@igormq any updates on model training?

@madhavajay thanks for the comments, model training time took a few days (I think 2-3 days total). I'll have a look at uploading the file temporarily however I hope we're not too far from completing the release!

**Comment by igormq** [comment]
I got 34.220 and 10.065  (WER and CER) at epoch 75 for the valid set.

**Comment by SeanNaren** [comment]
hmm that is a significant difference from the previous scores of our release. May need to run for more epochs, I can also check this out!

**Comment by igormq** [comment]
I do not think that it is training anymore, the loss value is already too low. Due to GPU mem constraint I reduced the batch size to 32

**Comment by buriy** [comment]
Hi @SeanNaren ,
1) you still count CER incorrectly in both test.py and train.py in V2: ( see #420 , whitespaces are counted incorrectly )
2) you still count CER and WER differently in test.py and train.py !
In train.py you calculate per-sample average ( https://github.com/SeanNaren/deepspeech.pytorch/blob/feature/V2/train.py#L318 ),
in test.py you calculate per-dataset average -- weighting results by sequence lengths! ( https://github.com/SeanNaren/deepspeech.pytorch/blob/feature/V2/test.py#L72 )
`A/C + B/D isn't equal to (A+C) / (B+D)` , right?

**Comment by SeanNaren** [comment]
@buriy definitely will address these before creating the release, thanks!

**Comment by buriy** [comment]
@SeanNaren also you have changed padding in around v1.2 -- I don't remember before or after, but old models can't be loaded because of that. New padding is better, because no more errors with too-short sequences, but old models can't be loaded -- maybe we could have a switch to allow both kinds of models, and so to be able to load old models and compare them?

**Comment by SeanNaren** [comment]
@buriy gah wish there was an easier way to support either, one reason for the new pre-trained models is to move away from the old padding... if there is enough support for having this capability or someones' willing to make a PR I'm all for it!

@igormq let's see what happens when I take @buriy changes and make the CER/WER calculations fair!

**Comment by buriy** [comment]
@SeanNaren yes I think after new models will be calculated old models won't be needed at all, so maybe there's no need to support both kinds of models after the V2 release -- only for fair checking them on the test sets (to have an apples-to-apples comparison). But I just copied model.py to model_old.py for that purpose.

**Comment by SeanNaren** [comment]
@buriy any chance you could look at #449?

**Comment by igormq** [comment]
@SeanNaren sure!

**Comment by buriy** [comment]
Hi Sean,
Looks fine, that will work, and I especially like the refactoring,
But I think macro averaging over all characters isn't needed, and just
plain average would work fine.
Plain average benefit is that it's additive: you can calculate scores over
several datasets and just average them.
So I think it's better.
And I don't remember whether simple averaging showed better score over
macro. Scientific papers are also usually silent about that (unless there's
a workshop with a shared quality measuring tool).
Maybe ask someone else, who knows it better? "Wer we are" author maybe?

пт, 2 авг. 2019 г., 20:04 Sean Naren <notifications@github.com>:

**Comment by miraodasilva** [comment]
Hi,

Any estimate on when this version will be released? Thanks a lot in advance!

Best regards,
Rodrigo Mira

**Comment by SeanNaren** [comment]
V2 is Released! https://github.com/SeanNaren/deepspeech.pytorch/releases/tag/v.2.0
